Uplift/Views/DropdownView.swift

```diff
@@ -0,0 +1,201 @@
+
+    init(headerView: UIView,
+         headerViewHeight: CGFloat,
+         contentView: UIView,
+         contentViewHeight: CGFloat,
+         halfDropdownEnabled: Bool = false,
+         halfOpenView: UIView? = nil,
+         halfOpenViewHeight: CGFloat = 0,
+         halfCloseView: UIView? = nil,
+         halfCloseViewHeight: CGFloat = 0,
+         halfHeight: CGFloat = 0) {
+        super.init(frame: .zero)
+
+        self.headerView = headerView
+        self.headerView.isUserInteractionEnabled = true
+        self.headerView.translatesAutoresizingMaskIntoConstraints = false
+        addSubview(self.headerView)
+
+        self.headerViewHeight = headerViewHeight
+        currentHeight = headerViewHeight
+
+        openCloseDropdownGesture = UITapGestureRecognizer(target: self, action: #selector(self.openCloseDropdown(sender:)))
+        self.headerView.addGestureRecognizer(openCloseDropdownGesture)
+
+        self.contentView = contentView
+        self.contentView.translatesAutoresizingMaskIntoConstraints = false
+        addSubview(self.contentView)
+
+        currentHeight = headerView.bounds.height
+        self.contentViewHeight = contentViewHeight
+        self.halfDropdownEnabled = halfDropdownEnabled
+
+        if let halfView = halfOpenView {
+            self.halfOpenView = halfView
+            halfView.translatesAutoresizingMaskIntoConstraints = false
+            halfView.isHidden = true
+            halfView.isUserInteractionEnabled = true
+            halfOpenDropdownGesture = UITapGestureRecognizer(target: self, action: #selector(self.openFromHalfDropdown(sender:)))
+            halfView.addGestureRecognizer(halfOpenDropdownGesture)
+            addSubview(halfView)
+        }
+
+        self.halfOpenViewHeight = halfOpenViewHeight
+
+        if let halfView = halfCloseView {
+            self.halfCloseView = halfView
+            halfView.translatesAutoresizingMaskIntoConstraints = false
+            halfView.isHidden = true
+            halfView.isUserInteractionEnabled = true
+            halfCloseDropdownGesture = UITapGestureRecognizer(target: self, action: #selector(self.closeHalfDropdown(sender:)))
+            halfView.addGestureRecognizer(halfCloseDropdownGesture)
+            addSubview(halfView)
+        }
+
+        self.halfCloseViewHeight = halfCloseViewHeight
+
+        self.halfHeight = halfHeight
+
+        setupConstraints()
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    func setupConstraints() {
+        headerView.snp.makeConstraints { make in
+            make.top.leading.trailing.equalToSuperview()
+            make.height.equalTo(headerViewHeight)
+        }
+
+        contentView.snp.makeConstraints { make in
+            make.top.equalTo(headerView.snp.bottom)
+            make.leading.trailing.equalToSuperview()
+            self.contentViewHeightConstraint = make.height.equalTo(0).constraint
```

Comment:
I've never seen this approach before of saving the contentView's height to a variable and updating it now and then. I think I would have used updateConstraints; just curious whether you considered updateConstraints to just change the height now and then and, if so, why did you choose this approach over it.

Reply:
SnapKit documentation describes both versions as ways to update a view's constraints. I decided to save a reference to the constraint rather than call `updateConstraints` as only one constraint is ever updated at a time, and this method uses fewer lines of code.